drivers/SmartThings/matter-sensor/src/init.lua

```diff
@@ -39,15 +39,53 @@ if version.api < 10 then
   clusters.RadonConcentrationMeasurement = require "RadonConcentrationMeasurement"
   clusters.TotalVolatileOrganicCompoundsConcentrationMeasurement = require "TotalVolatileOrganicCompoundsConcentrationMeasurement"
   clusters.SmokeCoAlarm = require "SmokeCoAlarm"
+  -- will require rain sensor and alarm capabilities
 end
 
+-- custom capability setup
+local rainSensorID = "smilevirtual57983.customRainSensor"
+local rainSensor = capabilities[rainSensorID]
+
 local BATTERY_CHECKED = "__battery_checked"
+local BOOLEAN_DEVICE_TYPES_CHECKED = "__boolean_device_types_checked"
 local TEMP_BOUND_RECEIVED = "__temp_bound_received"
 local TEMP_MIN = "__temp_min"
 local TEMP_MAX = "__temp_max"
 
 local HUE_MANUFACTURER_ID = 0x100B
 
+local BOOLEAN_DEVICE_TYPE_INFO = {
+  ["RAIN_SENSOR"] = {
+    id = 0x0044,
+    alarmComponentName = "rainSensorAlarm",
+  },
+  ["WATER_FREEZE_DETECTOR"] = {
+    id = 0x0043,
+    alarmComponentName = "waterFreezeDetectorAlarm",
+  },
+  ["WATER_LEAK_DETECTOR"] = {
+    id = 0x0041,
+    alarmComponentName = "waterLeakDetectorAlarm",
+  },
+  ["CONTACT_SENSOR"] = {
+    id = 0x0015,
+  }
+}
+
+local function set_device_type_per_endpoint(driver, device)
+  for _, ep in ipairs(device.endpoints) do
+      for _, dt in ipairs(ep.device_types) do
+          local this_id = dt.device_type_id
+          for name, info in pairs(BOOLEAN_DEVICE_TYPE_INFO) do
+              if this_id == info.id then
+                  device:set_field(name, ep.endpoint_id)
+              end
+          end
+      end
+  end
+  device:set_field(BOOLEAN_DEVICE_TYPES_CHECKED, 1, {persist = true})
+end
```

Comment:
I think you wanted to call this function only once after onboarding.
Can't driver call it only once at device_added function?

Reply:
Yep, it's only being called in device_init

Reply:
Specifically if it hasn't been done before, per the `BOOLEAN_DEVICE_TYPES_CHECKED` gate